### PR TITLE
Restrict PO box shipping

### PR DESCRIPTION
**Pull request by dkubb**
This branch will restrict PO Box shipping to just USPS and Canada Post.

Both Fedex and UPS cannot ship directly to a PO Box via their regular shipping methods. They both have special services that can ([SmartPost](http://www.fedex.com/us/smart-post/) and [SurePost](https://www.ups.com/content/us/en/resources/track/sp_definition.html) respectively), but those are not supported by ActiveShipping afaik.

### TODO

* [x] Add `Spree::AddressLine` to normalize and validate the address line
  * [x] Add `#pm_box?` to identify if an address line indicates a PO Box
* [x] Use `Spree::AddressLine` to setup the `ActiveShipping::Location` object
* [x] Add config setting that is tested to see if the carrier can ship to PO Boxes
  * [x] Base (`false`, default)
  * [x] USPS (`true`)
  * [x] Canada Post (`true`)

### CODE REVIEW

**Comment by dkubb** [comment]
@mbj can I get a 30% pass over this TODO and code?

**Comment by mbj** [comment]
> Add config setting that is tested to see if the carrier can ship to PO Boxes

Do we need a "config setting" when the values are the same in all environments. Also we thought about not upstreaming the change, so IMO we just want a query API with hardcoded values.

**Comment by mbj** [comment]
As we do not plan to upstream this change, what about moving most of the additions to the private repo? The `Spree::AddressLine` classes could be moved into `OurProject::AddressLine`.

**Comment by dkubb** [comment]
> As we do not plan to upstream this change

I actually do plan to upstream this. There was a closed issue in the upstream repo to add it, but then no-one followed up on it.

File: app/models/spree/address_line.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    id_number = /(?:\#\s?)?(?<id>\d+)/.freeze
+    id_alpha  = /(?<id>[A-Z])/i.freeze
+    id        = Regexp.union(id_number, id_alpha).freeze
+
+    POB_STRING = 'PO Box \k<id>'.freeze
+    PMB_STRING = 'PMB'.freeze
+    POB        = /\b#{po_box}#{separator}(?:#{id}\b)?/.freeze
+    PMB        = /\b#{PMB_STRING}\b/i.freeze
+
+    private_constant(*constants(false))
```
Comment (mbj) [inline app/models/spree/address_line.rb]: I wounder if we should move the `private_constant(*constants(false))` to the very bottom of the scope `end`. Or just after the last constant definition.

Actually I like the current location most, just want to make the "after last constant" an agreement of code style explicitly. I notice some of our projects are not consistent here.

Comment (dkubb) [inline app/models/spree/address_line.rb]: I like it just after all the constants are defined too. We should use this convention going forward.

File: app/models/spree/address_line.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
+module Spree
+  class AddressLine
+    separator = /\s*(?:\.\s*)?/i.freeze
+    post      = /P(?:[O0]ST)?/i.freeze
+    office    = /[O0](?:FFICE)?/i.freeze
+    box       = /(?:B(?:[O0]?X)?)?/i.freeze
+    po_box    = [post, office, box].join(separator.to_s).freeze
+    id_number = /(?:\#\s?)?(?<id>\d+)/.freeze
+    id_alpha  = /(?<id>[A-Z])/i.freeze
+    id        = Regexp.union(id_number, id_alpha).freeze
```
Comment (dkubb) [inline app/models/spree/address_line.rb]: @mbj these local variables look odd to me. I was thinking about changing them back to constants.

**Comment by mbj** [comment]
@dkubb What about from now on adding `adamantium` to freeze all freezable instances in code we touch, especially new classes we introduce? We have to start upstreaming this important invariant sooner than later. The more we upstream without the more we need to backfill / or painfully debug later.

File: app/models/spree/address_line.rb (hunk 2 above)
Comment (mbj) [inline app/models/spree/address_line.rb]: @dkubb What about to remove the `#freeze` calls? Than these look less odd, but we have the "lowest possible semantic unit used".

Comment (mbj) [inline app/models/spree/address_line.rb]: I mean while still kepping the `lvar` where the corresponding constant would not be accessed from methods.

**Comment by dkubb** [comment]
> What about from now on adding adamantium

We could join #spree on freenode.net and convince JDUtil of it's benefits. We could then get this merged into spree upstream on it's own, sneak it into the infrastructure, and then begin to use it in all commits we upstream.

File: app/models/spree/address_line.rb (hunk 1 above)
Comment (mbj) [inline app/models/spree/address_line.rb]: ack

File: app/models/spree/address_line.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    PMB        = /\b#{PMB_STRING}\b/i.freeze
+
+    private_constant(*constants(false))
+
+    def initialize(address_line)
+      @address_line = address_line.dup.tap do |line|
+        line.sub!(POB, POB_STRING)
+        line.sub!(PMB, PMB_STRING)
+        line.squish!
+        line.freeze
```
Comment (mbj) [inline app/models/spree/address_line.rb]: I'd prefer not to mutate, even when its a local variable that does not escape and do a traditional method chain:

```
line
  .sub(POB, POB_STRING)
  .sub(PMB, PMB_STRING)
  .squish
  .freeze
```

I'd only mutate intermediary in case a benchmark shows me its absolutely needed for performance. Also this method chaining form (which could also be done with mutating!) does not have redundant lvar `line` in the AST and is simpler from that POV.

**Comment by mbj** [comment]
@dkubb finished the requested 30% review, TODOs are fine IMO.

**Comment by mbj** [comment]
> We could join #spree on freenode.net and convince JDUtil of it's benefits. We could then get this merged into spree upstream on it's own, sneak it into the infrastructure, and then begin to use it in all commits we upstream.

Same for mutant, I want it at least as a dependency, initially not even part of the build. So I do not constantly have to rebase it in-and-out while upstreaming.

File: app/models/spree/address_line.rb (hunk 3 above)
Comment (dkubb) [inline app/models/spree/address_line.rb]: You're right. Plus the cost of the extra block probably dwarfs the string object allocation I was trying to save.

Comment (mbj) [inline app/models/spree/address_line.rb]: I'd generally not think to much about object allocation in ruby, and focus on correctness first. Going non mutating by default has the benefit that in case any (future) refactoring accidentally changes the "mutated value does not escape" - invariant: we are save.

Comment (mbj) [inline app/models/spree/address_line.rb]: BTW a sufficiently smart ruby runtime detects the "non escape" of the return value of non mutating string operations like `String#sub` and might turn this chain into an in-place mutation.

But I do not think ruby will ever reach this point. Maybe jruby supports this optimization in the future.

**Comment by dkubb** [comment]
@mbj this is ready for review.

File: app/models/spree/calculator/shipping/active_shipping/base.rb [hunk 4]
```diff
@@ -15,8 +15,19 @@ def self.service_name
           self.description
         end
 
+        def self.rate_from(result)
+          result[description]
+        end
+
+        def self.po_box_delivery?
+          false
```
Comment (mbj) [inline app/models/spree/calculator/shipping/active_shipping/base.rb]: you could define this body as: `self.class::PO_BOX_DELIVERY`. And have a constant in `Base` like `PO_BOX_DELIVERY = false`. Each subclass than can have a specific version of this `PO_BOX_DELIVERY` constant. Benefit: No need to test the method in the subclass anymore, as there is none. One test that shows the class relative constant lookup should be enough.

Comment (dkubb) [inline app/models/spree/calculator/shipping/active_shipping/base.rb]: Good idea. This is now fixed.

File: app/models/spree/calculator/shipping/active_shipping/base.rb [hunk 5]
```diff
@@ -233,18 +237,30 @@ def cache_key(package)
...
+        def build_location address
+          address_lines = %i[address1 address2]
+            .map(&address.method(:public_send))
+            .select(&:present?)
+            .map(&AddressLine.method(:new))
+          Location.new(
+            name:         address.try(:full_name) || address.try(:name),
+            company:      address.try(:company),
+            address1:     address_lines[0].to_s,
+            address2:     address_lines[1].to_s,
```
Comment (mbj) [inline app/models/spree/calculator/shipping/active_shipping/base.rb]: In case `address_lines` is guaranteed to be a two element array you might call `array_lines.first` and `array_lines.last` instead of having the index literal around. This is equivalent semantics with more primitive AST.

Comment (dkubb) [inline app/models/spree/calculator/shipping/active_shipping/base.rb]: I could use `#first` and `#second` instead. I don't want to use `#last` since it *could* break if the array gets a third entry.

Although, maybe this is even better:

```ruby
line1, line2 = %i[address1 address2]
  .map(&address.method(:public_send))
  .select(&:present?)
  .map(&AddressLine.method(:new))
```

Comment (dkubb) [inline app/models/spree/calculator/shipping/active_shipping/base.rb]: Oh wait, this won't work because I iterate over the list lower down. I will use `#second` instead.

File: app/models/spree/calculator/shipping/active_shipping/base.rb [hunk 6]
```diff
@@ -233,18 +237,30 @@ def cache_key(package)
           order = package.order
           ship_address = package.order.ship_address
           contents_hash = Digest::MD5.hexdigest(package.contents.map {|content_item| content_item.variant.id.to_s + "_" + content_item.quantity.to_s }.join("|"))
-          @cache_key = "#{stock_location}#{carrier.name}-#{order.number}-#{ship_address.country.iso}-#{fetch_best_state_from_address(ship_address)}-#{ship_address.city}-#{ship_address.zipcode}-#{contents_hash}-#{I18n.locale}".gsub(" ","")
+          @cache_key = "#{stock_location}#{carrier.name}-#{order.number}-#{ship_address.country.iso}-#{ship_address.state_text}-#{ship_address.city}-#{ship_address.zipcode}-#{contents_hash}-#{I18n.locale}".gsub(" ","")
         end
 
-        def fetch_best_state_from_address address
-          address.state ? address.state.abbr : address.state_name
+        def build_location address
```
Comment (mbj) [inline app/models/spree/calculator/shipping/active_shipping/base.rb]: You probably want to have parenthesis around the address argument.

File: app/models/spree/calculator/shipping/active_shipping/base.rb [hunk 7]
```diff
@@ -233,18 +237,30 @@ def cache_key(package)
           order = package.order
           ship_address = package.order.ship_address
           contents_hash = Digest::MD5.hexdigest(package.contents.map {|content_item| content_item.variant.id.to_s + "_" + content_item.quantity.to_s }.join("|"))
-          @cache_key = "#{stock_location}#{carrier.name}-#{order.number}-#{ship_address.country.iso}-#{fetch_best_state_from_address(ship_address)}-#{ship_address.city}-#{ship_address.zipcode}-#{contents_hash}-#{I18n.locale}".gsub(" ","")
+          @cache_key = "#{stock_location}#{carrier.name}-#{order.number}-#{ship_address.country.iso}-#{ship_address.state_text}-#{ship_address.city}-#{ship_address.zipcode}-#{contents_hash}-#{I18n.locale}".gsub(" ","")
```
Comment (mbj) [inline app/models/spree/calculator/shipping/active_shipping/base.rb]: I wounder if this code would be much easier to understand in case we'd build up an array of string components that form the key and than `join('-')` it together.

```
[
  stock_location,
  carrier_name,
  order.number,
  ...,
].join('-').gsub(' ', '')

**Comment by mbj** [comment]
@dkubb I'm done with the requested 90% review. I'll do a more verbose pass on the request for a final one.

File: app/models/spree/calculator/shipping/active_shipping/base.rb (hunk 7 above)
Comment (dkubb) [inline app/models/spree/calculator/shipping/active_shipping/base.rb]: I've cleaned up this method to be simpler.

**Comment by dkubb** [comment]
@mbj how far do you think I should go with mutation testing the modified pieces? I've tested that the `Spree::AddressLine` class is mutation tested.

The specs are pretty poor, and I think it would take an inordinate amount of time to bring them up to the level where things could be effectively mutation tested. At the same time, the code itself isn't so great, so I might end up having to either refactor significantly and/or use suboptimal integration specs to test the large monolithic methods.

I could constrain mutation testing to *just* the methods I added.

**Comment by mbj** [comment]
> I could constrain mutation testing to just the methods I added.

What about to also include the methods you changed, but there only kill the mutations in the lines you touched?

**Comment by dkubb** [comment]
@mbj I am going to mutation test all the code that I touched except:

* `Spree::Calculator::Shipping::ActiveShipping::Base#cache_key`
* `Spree::Calculator::Shipping::ActiveShipping::Base#compute_package`
* `Spree::Calculator::Shipping::ActiveShipping::Base#timing`

In all of these cases the change was a refactoring that simplified existing code. The first one has no public interface. The latter two were such minor changes that I'm confident that I didn't introduce any regressions -- the code and tests are so poor, or non-existent, I'd end up spending too much time rewriting them just to cover a couple line refactoring.

**Comment by dkubb** [comment]
I am verifying this with:

```bash
RAILS_ENV=test bundle exec mutant -r./spec/dummy/config/environment -j1 --use rspec -- \
  'Spree::AddressLine' \
  'Spree::Calculator::Shipping::ActiveShipping::Base#available?' \
  'Spree::Calculator::Shipping::ActiveShipping::Base.build_location' \
  'Spree::Calculator::Shipping::ActiveShipping::Base.deliver_to?' \
  'Spree::Calculator::Shipping::ActiveShipping::Base.rate_from' \
  'Spree::Calculator::Shipping::Usps::Base.rate_from'
```

**Comment by dkubb** [comment]
@mbj I am still running mutant against this, but I'm fairly certain it'll pass. In the meantime, this can probably go through a full code review.

File: app/models/spree/calculator/shipping/active_shipping/base.rb [hunk 8]
```diff
@@ -11,13 +11,48 @@ module ActiveShipping
...
+            name:         address.try(:full_name) || address.try(:name),
+            company:      address.try(:company),
+            address1:     address_lines.first.to_s,
+            address2:     address_lines.second.to_s,
+            country:      address.country.iso,
+            state:        address.state_text,
+            city:         address.city,
+            zip:          address.zipcode,
+            phone:        address.phone,
+            address_type: ('po_box' if address_lines.any?(&:po_box?))
```
Comment (mbj) [inline app/models/spree/calculator/shipping/active_shipping/base.rb]: Lets move this literal to a constant.

File: app/models/spree/calculator/shipping/active_shipping/base.rb [hunk 9]
```diff
@@ -11,13 +11,48 @@ module ActiveShipping
...
         def self.service_name
           self.description
         end
 
+        def self.rate_from(result)
+          result[service_name]
+        end
+
+        def self.build_location(address)
+          address_lines = %i[address1 address2]
```
Comment (mbj) [inline app/models/spree/calculator/shipping/active_shipping/base.rb]: What about to move the symbol array to a constant also?

File: app/models/spree/calculator/shipping/active_shipping/base.rb [hunk 10]
```diff
@@ -11,13 +11,48 @@ module ActiveShipping
...
+        def self.deliver_to?(location)
+          !location.po_box? || self::PO_BOX_DELIVERY
+        end
+
         def available?(package)
-          package.order.ship_address.present? && compute(package).present?
-        rescue Spree::ShippingError
+          ship_address = package.order.ship_address
+          ship_address.present? &&
+          self.class.deliver_to?(build_location(ship_address)) &&
```
Comment (mbj) [inline app/models/spree/calculator/shipping/active_shipping/base.rb]: I wounder if we should turn `deliver_to?` into an instance method, and making it private. I realize it would be a utility method, but there is no other call side using this method, so it could be specified through `#available?` fully.

File: app/models/spree/calculator/shipping/active_shipping/base.rb [hunk 11]
```diff
@@ -11,13 +11,48 @@ module ActiveShipping
...
+        PO_BOX_DELIVERY = false
+
+        delegate :build_location, to: self
+        private :build_location
+
         def self.service_name
           self.description
         end
 
+        def self.rate_from(result)
```
Comment (mbj) [inline app/models/spree/calculator/shipping/active_shipping/base.rb]: There is only one call side for this method, inside `#compute_session`, to reduce the public interface of the singleton lets move this to a private utility method. The semantics of this method (unlike for example `.build_location`) are so small that these can probably be easily specified via `#compute_session` specs.

File: app/models/spree/calculator/shipping/usps/base.rb [hunk 12]
```diff
@@ -2,30 +2,15 @@ module Spree
...
-          return nil if rates_result.kind_of?(Spree::ShippingError)
-          return nil if rates_result.empty?
-          rate = rates_result[self.class.service_code]
-
-          return nil unless rate
-          rate = rate.to_f + (Spree::ActiveShipping::Config[:handling_fee].to_f || 0.0)
-
-          # divide by 100 since active_shipping rates are expressed as cents
-          return rate/100.0
+        def self.rate_from(result)
```
Comment (mbj) [inline app/models/spree/calculator/shipping/usps/base.rb]: This method seems to be identical to the superclass implemenation, and can probably be dropped?

File: app/models/spree/calculator/shipping/active_shipping/base.rb (hunk 11 above)
Comment (dkubb) [inline app/models/spree/calculator/shipping/active_shipping/base.rb]: I found it extremely difficult to specify them via `#compute`. I don't understand it's specs at all, and I would prefer not to specify it as part of it.

I could move it to a private method still and not specify it directly or leave it as a singleton method.

File: app/models/spree/calculator/shipping/usps/base.rb (hunk 12 above)
Comment (dkubb) [inline app/models/spree/calculator/shipping/usps/base.rb]: No, it uses `service_code` and the other uses `service_name`. Why someone decided to name these different things, I couldn't tell you.

File: app/models/spree/calculator/shipping/active_shipping/base.rb (hunk 11 above)
Comment (mbj) [inline app/models/spree/calculator/shipping/active_shipping/base.rb]: > I found it extremely difficult to specify them via #compute. I don't understand it's specs at all, and I would prefer not to specify it as part of it.

Given the overall spagettiness of the code we have to touch here, this is sufficient argument for me to accept the current state.

Comment (dkubb) [inline app/models/spree/calculator/shipping/active_shipping/base.rb]: Given that these methods don't rely on instance state, or normal approach would be for them to be singleton methods. The fact that it made it possible to specify them without touching the crazily complex instance methods or their tests was a side benefit.

File: app/models/spree/calculator/shipping/usps/base.rb (hunk 12 above)
Comment (mbj) [inline app/models/spree/calculator/shipping/usps/base.rb]: wow, I did not spot this suptible difference.

File: app/models/spree/calculator/shipping/active_shipping/base.rb (hunk 11 above)
Comment (mbj) [inline app/models/spree/calculator/shipping/active_shipping/base.rb]: Fair enough.

File: app/models/spree/calculator/shipping/active_shipping/base.rb (hunk 10 above)
Comment (dkubb) [inline app/models/spree/calculator/shipping/active_shipping/base.rb]: Yeah, visually it looked a bit odd to have this here with other instance methods, but I'd argue those methods *should* have been singleton methods.

Instead of moving this I could define:

```ruby
delegate :deliver_to?, to: :class
```

This would allow it to remain as a singleton method, but make `#available?` internally consistent.

Comment (mbj) [inline app/models/spree/calculator/shipping/active_shipping/base.rb]: Alternative idea: What about to define `Address#location`? And move the whole body of `.build_location` there.

Comment (dkubb) [inline app/models/spree/calculator/shipping/active_shipping/base.rb]: `Spree::Address` isn't defined within spree_active_shipping unfortunately. We *could* monkey-patch it in, but I'd rather not do that.

Also there are two classes that `build_location` works with, `Spree::Address` and `Spree::StockLocation`. They have no common ancestor class or module.

**Comment by mbj** [comment]
Finished my code only pass. Given the pre-existing high noise level I have to admit its hard for me to squeeze out everything.

At least I think you did a good job in adding the po-box feature in a way it does not increase complexity in an exponential way.

**Comment by dkubb** [comment]
@mbj mutant passed with this code. It should be ready to merge.
